EIPS/eip-4844.md

```diff
@@ -239,30 +222,62 @@ def get_data_gasprice(header: Header) -> int:
     )
 ```
 
-The block validity conditions are modified to include data gas checks:
+The block validity conditions are modified to include data gas checks (see the [Execution-layer validation](#execution-layer-validation) section below).
+
+The actual `data_fee` as calculated via `calc_data_fee` is deducted from the sender balance before transaction execution and burned, and is not refunded in case of transaction failure.
+
+### Consensus-layer validation
+
+On the consensus-layer the blobs are referenced, but not fully encoded, in the beacon block body.
+Instead of embedding the full contents in the body, the blobs are propagated separately, as "sidecars".
+
+This "sidecar" design provides forward compatibility for further data increases by black-boxing `is_data_available()`:
+with full sharding `is_data_available()` can be replaced by data-availability-sampling (DAS) thus avoiding all blobs being downloaded by all beacon nodes on the network.
+
+Note that the consensus-layer is tasked with persisting the blobs for data availability, the execution-layer is not.
+
+The `ethereum/consensus-specs` repository defines the following consensus-layer changes involved in this EIP:
+
+- Beacon chain: process updated beacon blocks and ensure blobs are available.
+- P2P network: gossip and sync updated beacon block types and new blob sidecars.
+- Honest validator: produce beacon blocks with blobs; sign and publish the associated blob sidecars.
+
+### Execution-layer validation
+
+On the execution-layer, the block validity conditions are extended as follows:
 
 ```python
 def validate_block(block: Block) -> None:
     ...
 
-    num_blobs = 0
-    for tx in block.transactions:
+    block_data_gas = 0
+
+    blob_transactions = [tx for tx in block.transactions if type(tx) is SignedBlobTransaction]
+    for tx in blob_transactions:
```

Comment:
I think it is clear to the reader but really these checks should only apply to _just_ `SignedBlobTransaction`s and should be place on top of checks for existing transaction types

to illustrate what I mean, someone could read this as checking for "payability" twice, once during existing txn processing and again following line 267

the previous way it was written relies on some "type punning" from Python but it did respect this caveat

Reply:
I think going with 2 and adding branching within the loop might be the better option.

Reply:
I reverted the loop to the original version over all txs and added a payment logic section, then an if clause for all the specific blob tx logic. A bit more verbose, but I think this is the cleanest approach.